Add option to restore removed from cart

xavier-lc opened the pull request:
This PR contains the same changes as https://github.com/Automattic/wp-calypso/pull/103397, which was [reverted](https://github.com/Automattic/wp-calypso/pull/103469) in order to wait for translations to be ready. (see [comment](https://github.com/Automattic/wp-calypso/pull/103397#issuecomment-2883978172))

<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of CHE-29

## Proposed Changes

* When a user removes an item from the shopping cart, if there are still other items on it, add the option to restore the removed item. Removed items are tracked on the new `RestorableProductsContext`.
* In order to keep the item order in the cart when removed, the position in which they are rendered is tracked on the new `CartProductsOrderContext`.
* Reduce the number of products that have a removal confirmation modal, set on [the `getHasRemoveFromCartModal` function](https://github.com/Automattic/wp-calypso/pull/103397/files#diff-d6d276e55159d4aca386ac0391e990943d93b5dc197e6dc3fad4b0840db3f277R608)

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Hide whitespace to look at the diff:

<img src="https://github.com/user-attachments/assets/421d73cd-2a29-4322-919f-b15ab44726b9" width="250">

### Restore items

1. Open the live preview
2. Go to `/add-ons/{siteUrl}`
3. Under `Premium Themes`, click on `Buy add-on`
4. You'll be redirected to the checkout page
5. Go back to the add ons page, and repeat step 3 for `Custom CSS`
6. On the checkout page, you'll now have 2 items on the cart
7. On any of them, click on `Remove from cart`, and then `Continue` on the confirmation dialog
8. Check that a `Premium Themes was removed from the shopping cart. Restore` line appears
9. Click on `Restore`, check that the item is added back to your cart

![chrome-capture-2025-5-19](https://github.com/user-attachments/assets/56dd83f4-9fe2-4a5d-a9c6-b3972f5d8640)

### Restore items after changing product duration

1. Open the live preview
2. Go to `/start/domain/domain-only`
3. Search for a domain, select more than one domain, then click on `Continue`
4. Click on `Just buy domains`
5. Make changes on a product, then remove it
6. Make changes on the other products
7. Restore the removed product, check that it keeps the changes made on step 5

### Empty cart

Repeat the previous process, but now remove all items, and check that you are redirected out of the checkout page:

![chrome-capture-2025-5-14 (1)](https://github.com/user-attachments/assets/7e3ccd5b-ddba-44a1-b12c-e9b9c39f866d)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Commit:
Add `RestorableProductsProvider`

Commit:
Add `RemovedFromCartItems`

Commit:
Wrap `WPCheckoutMainContent` with `RestorableProductsProvider`

Commit:
Add `RemovedFromCartItems` on `WPCheckoutOrderReview`

Commit:
Handle restorable products on `CheckoutLineItem`

Commit:
Pass `restorableProducts`/` setRestorableProducts` to `LineItem`

Commit:
Handle `setRestorableProducts` promise after `removeProductFromCart` resolves

Commit:
Await on `addProductsToCart` before updating restorable products

Commit:
Return `state` object on `RestorableProductsProvider` so as to prevent re-renders

Commit:
Adjust `Restore` style

Commit:
Styling

Commit:
Disable button when restoring

Commit:
Show remove modal only when needed

Commit:
Update tests

Commit:
Adjust style according to design

Commit:
Add tests

Commit:
Merge branch 'trunk' into add/restore-removed-from-cart

Commit:
Merge branch 'trunk' into add/restore-removed-from-cart

Commit:
Add loading placeholder

Use the same as `LoadingContent` from `packages/composite-checkout/src/components/loading-content.tsx`

Commit:
Use core button

Commit:
Track cart product order when removing/restoring so the restore button option on the same place as the removed prod

Commit:
Extract duplicate code

Commit:
Fix order not taking into account plan duration change

Commit:
Adjust button color

Commit:
Add `isRestorable` prop

This prevents the restore behaviour from appearing on the masterbar minicart

Commit:
Add comment

Commit:
Rm unnecessary css

Commit:
Reset style for minicart

Commit:
Rename

fcoveram reviewed (approved):
It looks great 🚀

DavidRothstein reviewed (changes requested):
Impressive work overall.

I have not really reviewed the code, but I did a bit of testing and discovered a major problem.  If you add two domains to the cart at the same time, they get displayed on the left side of checkout with the same domain name, as shown below (and even weirder things happen if you try to remove one of them; it starts duplicating the cart items more):

![multiple-domains](https://github.com/user-attachments/assets/350eea26-79c8-432c-b8fc-fa2d77c56994)

As mentioned, I haven't really reviewed the code, but skimming through it I see that some things are keyed based on `product_id`, and that may be the source of the problem (both domains above have the same product ID since they're both `.com`).  I'm not sure the absolute correct way to distinguish between multiple instances of the same product in the cart, but using a combination of `product_id` and `meta` might go pretty far.

Separately (but not as high priority), the change to the "Remove from cart" styling seems to make it quite prominent -- it looks almost as prominent as the name of the product you're buying:

**Before:**
![before](https://github.com/user-attachments/assets/04c1139e-2e6a-4360-995e-e931abcdf3b7)

**After:**
![after](https://github.com/user-attachments/assets/51282d86-1d70-4e1a-8e42-3d36e41705fa)

That seems like a decent-size design change (which also makes this link look different and more prominent than other links within checkout).  Do we really need to change the design of that link at all as part of this issue?

sirbrillig commented:
> if you add two domains to the cart at the same time, they get displayed on the left side of checkout with the same domain name,

Ugh, I didn't notice that. I think that changed since I last reviewed the code. I see lots of uses of the product ID as a key which definitely won't work.

>  I'm not sure the absolute correct way to distinguish between multiple instances of the same product in the cart, but using a combination of product_id and meta might go pretty far.

`product.uuid` is probably a safe bet unless it needs to be persistent across cart changes, in which case it depends a lot on what you mean by "the same product" (does the product with a different volume, billing term length, quantity, or price mean it's the same or different?).

I also haven't read through the code closely enough to see why the items are being grouped but we really shouldn't do that. I'll see if I can offer more specific comments inline.

sirbrillig commented on client/my-sites/checkout/src/components/wp-order-review-line-items.tsx:
```diff
@@ -127,6 +139,22 @@ export function WPOrderReviewLineItems( {
+			setCartProductsOrder( responseCart.products.map( ( { product_id } ) => product_id ) );
+		}
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps -- Having cartProductsOrder as a dependency would trigger an infinite loop.
```
You can use a callback to `setCartProductsOrder()` instead of using `cartProductsOrder` directly. The callback will receive the current state as an argument. That will be more reliable and not require this hack. You can put the `productOrderNeedsInitialization || cartHasBeenUpdated` conditions inside that callback.

xavier-lc replied:
Thanks for the suggestion, much better this way!

sirbrillig commented on client/my-sites/checkout/src/components/wp-order-review-line-items.tsx:
```diff
@@ -190,39 +218,60 @@ export function WPOrderReviewLineItems( {
 	);
 
+	const cartItems = responseCart.products.reduce< LineItemByProductId >( ( acc, product ) => {
+		acc[ product.product_id ] = (
```
As noted in the main comment thread, `product_id` is not unique to a cart item; you can have multiple of the same product in the cart at the same time. `product.uuid` is more reliable but will change when the cart is modified or reloaded so it may not be the best for your purposes either.

🤔 Store shopping cart items on the backend do have a `get_id_for_serialization()` method which is used for certain caching operations and could easily be exposed through the shopping cart endpoint (see `get_cart_item_for_js_cart()`. That would be unique and persist across cart changes, unless some property of that cart item changes, which may be helpful for you here. In fact, we could use it to set `uuid` [instead of just an integer](https://github.com/Automattic/wp-calypso/blob/83f75f09b5e0eacbcbb0ea36d014a4cd63d96265/packages/shopping-cart/src/cart-functions.ts#L252). I had always planned on doing that at some point.

xavier-lc replied:
@sirbrillig 

I've tried adding `'item_id' => $cart_item->get_id_for_serialization()` on `get_cart_item_for_js_cart`, and then using it to set the `uuid` on `convertRawResponseCartToResponseCart`, but I'm finding that the serialization value changes depending on all the products in the cart. For example, if I have a business plan, and 2 domains and remove one of the domains, then the serialization of the plan and the remaining domain changes. So I don't think I can rely on that for the ordering.

I'm thinking that I could try to get a hash value on the client with a mix of the `product` object's values, but I imagine that could end up breaking up unexpectedly as well, so I'm leaning towards listing the removed products by themselves unless you have any other ideas to keep the order?

Thanks!

sirbrillig replied:
> For example, if I have a business plan, and 2 domains and remove one of the domains, then the serialization of the plan and the remaining domain changes. So I don't think I can rely on that for the ordering.

Hm... I wonder how that's happening? Anyway, you're right; the ID for serialization includes a lot of pricing data and is designed to err on the side of caution so that we regenerate prices if anything changes.

I made a PR to provide hopefully a more stable value: 183564-ghe-Automattic/wpcom Let me know what you think!

xavier-lc replied:
> I made a PR to provide hopefully a more stable value: 183564-ghe-Automattic/wpcom Let me know what you think!

That works great, thanks!

sirbrillig replied:
Cool. Will merge that shortly.

xavier-lc commented:
> As mentioned, I haven't really reviewed the code, but skimming through it I see that some things are keyed based on `product_id`, and that may be the source of the problem (both domains above have the same product ID since they're both `.com`). I'm not sure the absolute correct way to distinguish between multiple instances of the same product in the cart, but using a combination of `product_id` and `meta` might go pretty far.

Good catch! I'm using the product ids to track the product order so that when the product is removed, the restore option can be shown on the same place. 

> product.uuid is probably a safe bet unless it needs to be persistent across cart changes, in which case it depends a lot on what you mean by "the same product" (does the product with a different volume, billing term length, quantity, or price mean it's the same or different?).

I tried that, but when the product was restored the `uuid` changed :/

Commit:
Simplify `useEffect` hook with state setter callback

fcoveram commented:
To @DavidRothstein 

> Separately (but not as high priority), the change to the "Remove from cart" styling seems to make it quite prominent -- it looks almost as prominent as the name of the product you're buying:
> 
> Before:
> 
> <img src="https://private-user-images.githubusercontent.com/235183/446211745-04c1139e-2e6a-4360-995e-e931abcdf3b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc5MTAxOTcsIm5iZiI6MTc0NzkwOTg5NywicGF0aCI6Ii8yMzUxODMvNDQ2MjExNzQ1LTA0YzExMzllLTJlNmEtNDM2MC05OTVlLWU5MzFhYmNkZjNiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyMlQxMDMxMzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMDUxMDQ3MzhmMmRiZTEzNDY4NWRiOTEyZTJjNTVjYWE3YzkwMjZkYjhmMzZhNDdmYzE0YTA5MmIzNjVlYWNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bHpK2I-GS178_JppZNgfOJz_Vo51x5SpXIKqfwknHi0" alt="image" width=600>
> 
> After:
> 
> <img src="https://private-user-images.githubusercontent.com/235183/446211742-51282d86-1d70-4e1a-8e42-3d36e41705fa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc5MTAxOTcsIm5iZiI6MTc0NzkwOTg5NywicGF0aCI6Ii8yMzUxODMvNDQ2MjExNzQyLTUxMjgyZDg2LTFkNzAtNGUxYS04ZTQyLTNkMzZlNDE3MDVmYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyMlQxMDMxMzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZWJmZmEyNzNmNGI4YTU1YThkOTUyZmVhNmFiMmIwZjVlZmI2MjkzYjUxZWUzMTQxMzIxNDQ4YWM5NGFhNzRlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ohaRWWkcdMpoMIVetfc0tTG7Y5AyFJlKxzvKV1pV-bo" alt="" width=600>
> 
> That seems like a decent-size design change (which also makes this link look different and more prominent than other links within checkout). Do we really need to change the design of that link at all as part of this issue?

This change was intentional. Part of the design updates in the onboarding included making the links more predominant as these are secondary/tertiary actions across the flows. The consistency pass involved making all these actions to look equal by using the same component, hopefully the Core one, and apply the less changes possible.

Commit:
Use new backend field `cart_item_id` as product `uuid`

Commit:
Restore/remove while keeping track of the position and product modifications working

Just need to check the scenario where a product is added to the cart from outside the same page (e.g. a different tab)

Commit:
Extract initialize to a separate hook

DavidRothstein commented:
@fcoveram Checkout is frequently accessed outside of onboarding, so it doesn't always have the same design considerations.  But my bigger point was this:
- Why would we change this for the "Remove from cart" link only, and not other links in checkout?  It introduces an inconsistency (see below) -- not that the link presentation in checkout is amazing currently, but at least there's some consistency where links we want to emphasize are blue and links we don't want to emphasize are light gray.  With this pull request, we're now introducing a third link style, with no clear reason.
- Why would we do it as part of this pull request?  This kind of change really should be discussed in its own issue.

![checkout-links](https://github.com/user-attachments/assets/abb94b19-e287-4ec4-bdc8-12ba553b49eb)

Commit:
Rm comments

Commit:
Merge branch 'trunk' into add/restore-removed-from-cart

Commit:
Add `cart_item_id` on `getEmptyResponseCartProduct`

Commit:
Take only active prods into account for modifications

Commit:
Update

Commit:
Add `cart_item_id` to fixtures so tests don't break

sirbrillig commented on client/my-sites/checkout/src/components/wp-order-review-line-items.tsx:
```diff
@@ -127,6 +139,68 @@ export function WPOrderReviewLineItems( {
+				( uuid ) => ! prevCartProductsOrderActiveUuids.includes( uuid )
+			);
+
+			if ( hasNewUuid ) {
```
What would happen if a user changed the cart via some means other than the remove/restore mechanic or the term variant picker? 

For example, IIRC there's sometimes an upsell in the checkout sidebar that can add a plan to your cart if the cart contains only a domain (that might have changed but it's definitely possible and should be supported).

Another example – and one that's easier to test – is if the cart changes outside of checkout: the shopping cart reloads on refocus, so if you were to open a new tab, modify the cart to add a product or remove a product, then refocus checkout, the cart could be totally different than it was previously, but I think this caching would not be updated.

I think you might be able to expand the logic in this effect to compare more than one product and also compare adding and removing separately which might resolve this.

xavier-lc replied:
> For example, IIRC there's sometimes an upsell in the checkout sidebar that can add a plan to your cart if the cart contains only a domain (that might have changed but it's definitely possible and should be supported).

Ah, I've never seen it, do you have in mind any related PR that could help me replicate that?

> Another example – and one that's easier to test – is if the cart changes outside of checkout: the shopping cart reloads on refocus, so if you were to open a new tab, modify the cart to add a product or remove a product, then refocus checkout, the cart could be totally different than it was previously, but I think this caching would not be updated.

I've tried taking that into account as well, but the fact that `responseCart.products` is updated before `restorableProducts` prevents me from being able to check whether the product has been removed in the page of outside of it. I'd rather try to check whether the cart has been updated externally (not sure how yet), and if that's the case accept the new data as the source of truth and reset the restorable products to empty. Wdyt?

sirbrillig replied:
I came up with a testable scenario where this issue manifests. It's a bit contrived but I think there are others which have the same result.

1. Add a plan and a domain to your cart and visit checkout.
2. Click to remove the domain from your cart. Verify that you see the "Domain registration was removed..." action item in its place.
3. Open another tab and visit `/domains/add` for the same site.
4. In the second tab, click to add a _different_ domain to your cart.
5. Wait about 1 minute to pass the refresh-on-focus timeout.
6. Return to the first tab with checkout _without_ reloading.

In production, you'll see the cart refresh and the new domain appear in the list.

In this PR you'll see the cart refresh but the new domain will not appear in the main column, even though it does appear in the sidebar and the price is included.

<img width="1076" alt="Screenshot 2025-05-29 at 3 52 16 PM" src="https://github.com/user-attachments/assets/34497daa-58f4-4909-9b2a-0873e130f990" />

sirbrillig replied:
>  I'd rather try to check whether the cart has been updated externally (not sure how yet), and if that's the case accept the new data as the source of truth and reset the restorable products to empty. Wdyt?

I think that's a good idea, but the difficulty is knowing what "updated externally" means. The shopping cart only has one kind of update; a requested change is made, then we do a round-trip to the server and replace the existing cart object with a new one. I believe you mean that you want to know if the cart is updated in a way that doesn't keep some percentage of the same products, so I think the only way to do that is to compare the `uuid` of the products after an update to what they were previously. But since they will always differ when any product is removed, it's hard to say what the threshold is for resetting the state.

This is a tricky problem to resolve. I'm brainstorming some things, but I have no solid ideas just yet. I think maybe we don't need multiple useEffect calls and that just one will do. I think the key is to always render the line items that are currently in the cart, no matter what the cached uuids are.

xavier-lc replied:
> I think the key is to always render the line items that are currently in the cart, no matter what the cached uuids are.

Right, that makes sense 👌

xavier-lc replied:
@sirbrillig I've ended up dropping the ordering because that was what was making the rendering unnecessarily tricky

sirbrillig replied:
That's reasonable. I think it's possible to do, and I think you were on the right track, but the complexity is hard to get right. Perhaps that can be a future addition.

DavidRothstein replied:
>For example, IIRC there's sometimes an upsell in the checkout sidebar that can add a plan to your cart if the cart contains only a domain (that might have changed but it's definitely possible and should be supported).

For future reference (and in case we ever try adding the ordering feature back) the above is definitely a relevant example and very easy to test:
1. Start with a free site that already exists (i.e. has already gone through onboarding).
2. Put a domain in the shopping cart.
3. Click the upsell link in the sidebar (which will add a Personal plan to the cart).

Good find with the overall issue here!

sirbrillig reviewed (commented):
This is looking much better! Thank you for the changes.

I have one more concern about the caching of the cart order data when the cart changes for reasons we don't expect.

xavier-lc committed:
Merge branch 'trunk' into add/restore-removed-from-cart

Commit:
Handle products added on a different tab

Commit:
Drop keeping the product order when restoring/removing

Rendering the actual products from our order instead of from the cart response gets troublesome if we need to take into account changes made to the cart from a different tab.

sirbrillig reviewed (approved):
I believe this addresses all the issues we saw. Thank you for your persistence!

DavidRothstein commented:
>Designs (i3): [W9xI27S6Swvw5Ku21EbZvn-fi-11330_20912](https://href.li/?https://www.figma.com/file/W9xI27S6Swvw5Ku21EbZvn/?node-id=11330-20912)

Thanks @fcoveram -- that's helpful!

To me, that illustrates the value of a separate issue to carefully implement the designs, because some of the feedback I had above about the implementation in this pull request does not seem as relevant when compared to the designs as a whole.  I think this pull request just seems to leave things in an odd in-between state.

Can we consider removing the styling changes to the existing links from this pull request before merging it?

>I can open tickets for each of the changes applied in the page and revisit there their feasibility.

Good idea.  I can do one for the link styles specifically now.

DavidRothstein commented:
>Good idea. I can do one for the link styles specifically now.

https://linear.app/a8c/issue/CHE-163/checkout-implement-new-links-styles

fcoveram commented:
> Can we consider removing the styling changes to the existing links from this pull request before merging it?

Sure! I will create Linear issues for each of the changes. Thanks

Commit:
Undo style changes to `Remove from cart` button

Commit:
Merge branch 'trunk' into add/restore-removed-from-cart

Commit:
Undo style changes to `Remove from cart` button

fcoveram commented:
<!-- linear:replyTo:2881777466 -->

I just checked, and it looks good ✨

DavidRothstein reviewed (approved):
Looks good to me -- seems ready to go :+1:

Commit:
Merge branch 'trunk' into add/restore-removed-from-cart
